### PR TITLE
[junit-platform tester] Allow quoted test names

### PR DESCRIPTION
**Pull request by kriegfrj**
This PR is a partial fix for #3580.

### CODE REVIEW

**Comment by bjhargrave** [comment]
>     3\. Do we need to do anything to the gradle or maven plugins for quoting test names?

I think the answer here is that the ProjectTesterImpls need to quote, if necessary, each element in the generated `tester.names` property.

**Comment by kriegfrj** [comment]
> > 
> > 3\. Do we need to do anything to the gradle or maven plugins for quoting test names?
> > 
> 
> I think the answer here is that the ProjectTesterImpls need to quote, if necessary, each element in the generated `tester.names` property.

Hmm, I think that's part of the answer. However I've started looking at this and I think it's worth pausing to take stock and quickly looking at all of the consumers to figure out an implementation that is consistent. There are a few different players here:

1. JUnit 3, 4 & 5
2. maven-surefire-plugin and maven-failsafe-plugin
3. Gradle's built-in test task
4. The old Bnd tester
5. Bndtools
6. Bnd gradle plugin
7. Bnd maven tester plugin

===JUnit===

* JUnit Platform syntax is: `<fully-qualified classname>#<method name>(parameter1Type, parameter2Type)`
* The bit in parentheses is required for parameterized methods, but optional for methods that don't take parameters.
* There is no easy way to filter based on a certain test iteration for a parameterized test.

===Maven plugins===

Documentation is sparse (https://maven.apache.org/surefire/maven-surefire-plugin/examples/single-test.html and https://maven.apache.org/surefire/maven-failsafe-plugin/examples/single-test.html), but:

* They use syntax -Dtest=<filter> (-Dit.test for failsafe plugin).
* `#` is used as the method name delimiter. Thus, syntax is loosely based on JUnit.
* There doesn't appear to be a way to specify a particular method if it is overloaded with different parameter counts.
* Individual parameter iterations can be filtered using [] syntax. Filter is based on the display name of the test.
* What documentation exists is for JUnit 4 parameterized tests. I"m not sure to what extent the above applies to JUnit 5. However, I have verified empirically that JUnit 5 parameterized tests work, but the [] filter syntax does not.

===Gradle===

Appears to be two syntacies - one similar to maven-surefire (-D), and the `--tests` command line. The latter seems to be more recent.

The syntax seems to be similar to the maven plugins, except that the method delimiter is "." rather than "#". Given the inherent ambiguity arising from this (as . also delimits packages and classes), they also seem to use the convention that the test classes starts with capital letters.

https://docs.gradle.org/current/userguide/java_testing.html#test_filtering

===Old Bnd tester===

* Exact matches only, using FQN of class
* Uses ":" as the method delimiter for `tester.names`.
* No support for parameterized tests.

===Bndtools===

* Does whatever the tester does.
* Not sure how well it supports parameterized test methods out-of-the-box.
* We have the most flexibility here because it's all handled under-the-hood.

===Bnd maven and gradle testing plugins===

* Seem to be caught between emulating the syntax of their respective environment, or simply using Bnd tester syntax.
* Eg, @rotty3000 was using the `-Dtest=` of maven, but the `:` delimiter of the Bnd tester syntax.

So we need to make a decision here about whether these plugins try to emulate the maven/gradle syntax, or else whether they simply use the Bnd tester syntax.

Thoughts?

**Comment by bjhargrave** [comment]
For OSGi testings (vs normal JUnit testing), all test execution flows through 

https://github.com/bndtools/bnd/blob/2171d1a6462402843e290284919d3c8ab11c2b7b/biz.aQute.bndlib/src/aQute/bnd/build/Project.java#L2356

which takes a `List<String>` of test names. 

Gradle:

https://github.com/bndtools/bnd/blob/2171d1a6462402843e290284919d3c8ab11c2b7b/biz.aQute.bnd.gradle/src/aQute/bnd/gradle/TestOSGi.groovy#L94

Maven:

https://github.com/bndtools/bnd/blob/2171d1a6462402843e290284919d3c8ab11c2b7b/maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java#L188

Bndtools uses ProjectTester.addTest directly:

https://github.com/bndtools/bnd/blob/2171d1a6462402843e290284919d3c8ab11c2b7b/bndtools.core/src/bndtools/launch/OSGiJUnitLaunchDelegate.java#L305-L321

This `List<String>` of test names is converted into the `tester.names` property by the ProjectTester impls

https://github.com/bndtools/bnd/blob/2171d1a6462402843e290284919d3c8ab11c2b7b/biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/plugin/ProjectTesterImpl.java#L68-L72

https://github.com/bndtools/bnd/blob/2171d1a6462402843e290284919d3c8ab11c2b7b/biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java#L57-L61

So this is where any quoting should be done if necessary since this is where the list is flattened to a property value.

Each test activator (`aQute.junit` or `aQute.tester.junit.platform`) should then be responsible for  parsing the property back into a list and mapping the value into the form expected by the respective testing engine.

So the form in the `tester.names` property should not need any changing other than to support quoting elements if the element uses a `,` in parameters.

We can change Bndtools to provide more precise element values since it should know the full parameters of the selection in the UI.

Gradle and Maven users can supply more precise forms on the command line invocation which will be passed into the ProjectTester.addTest method.

So, in summary, the ProjectTester impls need to quote elements, if necessary, when assembling the `tester.names` property value. The test activators need to handle quoted elements when parsing the `tester.names` property value back into a list. The test activators need to handle any mapping to the forms expected by there respective test engines.

**Comment by rotty3000** [comment]
Speaking mostly to the maven scenario, in an ideal world, bnd would be able to run it's tests within `maven-failsafe-plugin` using its cli interface. However, we also don't want to spend an inordinate amount of time nor be burdened with baggage just to make that work. So, barring the direct use of `maven-failsafe-plugin` we'd try to emulate it as closely as possible in order to make developers comfortable.

So, I think the overarching goal here is that while we empower developers with the best possible OSGi testing experience we also make them comfortable. Whatever _that_ is is what we should do. Oh, and let's also be practical.

**Comment by kriegfrj** [comment]
Sorry, I was in a bit of a rush when I posted that comparison of syntacies yesterday, so perhaps the main point I was trying to make got lost. 

@bjhargrave , I definitely agree that the `ProjectTesterImpl` is the best place to quote the elements of `tester.names`. I'm more concerned about divergence other parts of the `tester.names` syntax. Like what @rotty3000 pointed out here:

> Speaking mostly to the maven scenario, in an ideal world, bnd would be able to run it's tests within `maven-failsafe-plugin` using its cli interface. However, we also don't want to spend an inordinate amount of time nor be burdened with baggage just to make that work. So, barring the direct use of `maven-failsafe-plugin` we'd try to emulate it as closely as possible in order to make developers comfortable.
> 

At the moment, as @bjhargrave pointed out here https://github.com/bndtools/bnd/issues/3580#issuecomment-557142977, the gradle and maven plugins simply pass in the `-Dtest=` / `--tests` parameters directly to the tester unmodified.

I think this is what I was getting at. The current syntax used by the Bnd testers for `tester.names` is incompatible with every other syntax - maven-failsafe/surefire-plugin, Gradle, and JUnit 5. `BundleEngine` uses JUnit 5 internally (method delimiter = `#`) and the junit.platform.Activator translates from the Bnd tester syntax (method delimiter = `:`) to JUnit 5 syntax. I'm questioning the value of maintaining a fourth syntax that is incompatible with everything else. I think it would be better if we simply adopted JUnit 5 syntax for the elements of `tester.names`. If necessary/convenient, we should also get the Maven/Gradle plugin wrappers to translate from maven/Gradle syntax to the tester syntax.

Perhaps this part of the discussion should be split into a different ticket and leave this one to deal with the quoting.

**Comment by bjhargrave** [comment]
> fourth syntax

What are all these syntax's?

For Bnd's Gradle TestOSGi task, `--tests` uses the normal Bnd `tester.names` format which is `class:method`. The format Gradle uses for plain JUnit testing to complicated with wildcard matching and we don't attempt to support it for OSGi testing. The reason we cannot support doing anything with wildcards is that we cannot map them to concrete test names in the gradle plugin as the plugin does not know what the execution runtime looks like. The TestOSGi task is just a simple driver of Bnd and we should not teach it to duplicate Bnd function.

For Bnd's Maven bnd-testing-maven-plugin, it also uses the normal Bnd `tester.names` format which is `class:method`. The maven-failsafe-plugin uses a complicated format supporting wildcard which I also don't think we can support. I am not even sure why we are discussing maven-failsafe-plugin here since we have bnd-testing-maven-plugin and no plan or understanding of how maven-failsafe-plugin could replace it for OSGi testing. Like as in the Bnd gradle support, the maven plugin is also a simple driver of Bnd. It also does not know about the execution environment to turn wildcards into concrete test names.

So for both our gradle and maven support we use the original Bnd `tester.names` format of `class:method` which we should continue to use for compatibility reasons. We should, however, expand `tester.names` format to allow parameters so that Bndtools can pass more precise information. `class:method(type1,type2)`. This format extension requires we support quoting elements in the `tester.names` property value so we can parse it correctly in the junit activator side.

So I see one format, extended to support parameters, here.

Now, at the engine level (aQute.junit, aQute.tester.junit.platform), we need to map this one syntax into a form the engine understands. The `replace(':',"#')` in junit-platform is basically an internal implementation choice since ultimately, we must supply DiscoverySelectors to the engine. How we map the Bnd `tester.names` format into DiscoverySelectors is our internal implementation problem.

But externally, the Bnd `tester.names` format has one syntax we plan to extend for method parameter specification. And that syntax is what Gradle users will put in `--tests` and Maven users will put in `-Dtest`. So nothing needs to be done in the maven or gradle plugins. They just pass the collected test values into the Project.test method in a `List<String>` (using the Bnd `tester.names` format) and we take it from there.

If we think we want to support some form of wildcard in the maven and gradle plugins, then we need to add that to the ONE Bnd `tester.names` format so that there is consistency and also so that the activators who are IN the runtime can map the wildcards to concrete tests.

**Comment by kriegfrj** [comment]
To answer your question: why are we discussing maven-failsafe/surefire-plugin, I'll quote @rotty3000 :  "barring the direct use of maven-failsafe-plugin we'd try to emulate it as closely as possible in order to make developers comfortable." In other words, to reduce the learning curve, the bnd-tester-maven-plugin should try and implement the same syntax if possible, or at least as similar as possible. Ditto for Gradle --tests support.

What I'm getting from your post is that this is impractical, partly because you need the OSGi environment to properly scan for the tests, and the maven/gradle wrappers are responsible for launching that environment. Makes sense.

I thought perhaps we could do it in a limited form - eg, support "#" for Maven, "." for Gradle. This could be done either by supporting these alternatives natively in the testers' format, or else having the maven/gradle plugins perform the translation to the native testers' format. This should be much easier that supporting all of the wildcard syntax (though I think that wildcards and tag support is something we should consider at some point too).

File: biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java [hunk 1]
```diff
@@ -29,6 +30,10 @@ public ProjectTesterImpl(Project project) throws Exception {
 		super(project);
 	}
 
+	static String maybeQuote(String s) {
+		return (s.indexOf(',') == -1) ? s : '"' + s + '"';
```
Comment (bjhargrave) [inline biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: It is a bit more complicated than this. The user may have already quoted the string! `--tests="class:method(int,int)"`.

See https://github.com/bndtools/bnd/blob/593fc85bbb09dd6b4c1e65ae4637142987e2019a/aQute.libg/src/aQute/libg/qtokens/QuotedTokenizer.java#L222-L250 for a robust quoting example.

File: biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/plugin/ProjectTesterImpl.java [hunk 2]
```diff
@@ -34,6 +35,10 @@ public ProjectTesterImpl(Project project, Container me) throws Exception {
 		this.me = me;
 	}
 
+	static String maybeQuote(String s) {
+		return (s.indexOf(',') == -1) ? s : '"' + s + '"';
```
Comment (bjhargrave) [inline biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/plugin/ProjectTesterImpl.java]: Need to handle better as above.

File: biz.aQute.tester/src/aQute/tester/plugin/ProjectTesterImpl.java [hunk 3]
```diff
@@ -32,6 +33,10 @@ public ProjectTesterImpl(Project project, Container me) throws Exception {
 		this.me = me;
 	}
 
+	static String maybeQuote(String s) {
+		return (s.indexOf(',') == -1) ? s : '"' + s + '"';
```
Comment (bjhargrave) [inline biz.aQute.tester/src/aQute/tester/plugin/ProjectTesterImpl.java]: Need to handle better as above.

File: biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java [hunk 4]
```diff
@@ -57,7 +62,9 @@ public boolean prepare() throws Exception {
 			Collection<String> testnames = getTests();
 			if (testnames.size() > 0) {
 				launcher.getRunProperties()
-					.put(TESTER_NAMES, Processor.join(testnames));
+					.put(TESTER_NAMES, Processor.join(testnames.stream()
```
Comment (bjhargrave) [inline biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: It seems busy to stream/map the list into a list and then join into a string. We can go straight there

```java
.put(TESTER_NAMES, testnames.stream()
  .map(ProjectTesterImpl::maybeQuote)
  .collect(Collectors.joining(",")));
```

File: biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/plugin/ProjectTesterImpl.java [hunk 5]
```diff
@@ -68,7 +73,9 @@ public boolean prepare() throws Exception {
 			Collection<String> testnames = getTests();
 			if (testnames.size() > 0) {
 				launcher.getRunProperties()
-					.put(TESTER_NAMES, Processor.join(testnames));
+					.put(TESTER_NAMES, Processor.join(testnames.stream()
```
Comment (bjhargrave) [inline biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/plugin/ProjectTesterImpl.java]: It seems busy to stream/map the list into a list and then join into a string. We can go straight there

```java
.put(TESTER_NAMES, testnames.stream()
  .map(ProjectTesterImpl::maybeQuote)
  .collect(Collectors.joining(",")));
```

File: biz.aQute.tester/src/aQute/tester/plugin/ProjectTesterImpl.java [hunk 6]
```diff
@@ -60,7 +65,9 @@ public boolean prepare() throws Exception {
 			Collection<String> testnames = getTests();
 			if (testnames.size() > 0) {
 				launcher.getRunProperties()
-					.put(TESTER_NAMES, Processor.join(testnames));
+					.put(TESTER_NAMES, Processor.join(testnames.stream()
```
Comment (bjhargrave) [inline biz.aQute.tester/src/aQute/tester/plugin/ProjectTesterImpl.java]: It seems busy to stream/map the list into a list and then join into a string. We can go straight there

```java
.put(TESTER_NAMES, testnames.stream()
  .map(ProjectTesterImpl::maybeQuote)
  .collect(Collectors.joining(",")));
```

File: biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java (hunk 4 above)
Comment (kriegfrj) [reply biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: :+1: wasn't sure if `Processor.join()` was doing some other magic.

File: biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java (hunk 1 above)
Comment (kriegfrj) [reply biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: Wouldn't it be better if the process responsible for populating testernames ensure that the strings that go into the list are normalized to begin with?

Comment (bjhargrave) [reply biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: This means that every driver must do this instead of in fewer places. Better to do this at the choke points.

Comment (kriegfrj) [reply biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: Better to do this at the choke points, yes, but I think that the choke point is where the model is populated (`ProjectTester.addTest()`). That way, the drivers don't have to worry about it either (as they all use `addTest()`). If I do it in `ProjectTesterImpl` I have to do it three times, plus any other third-party consumers of `getTests()` will also have to handle it.

Comment (kriegfrj) [reply biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java]: I have added code to strip quotes (if present) in `ProjectTester.addTest()`.
